### PR TITLE
Fixed-index slicing overloads for 3D arrays

### PR DESCRIPTION
**Pull request by nelson-wu**
For the first half of https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1077-3d-4d-fixed-index-slicing.md

### CODE REVIEW

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 1]
```diff
@@ -5409,13 +5517,6 @@ namespace Microsoft.FSharp.Core
            when ^T : byte        = RangeByte    (retype start) (retype step) (retype finish)
         
 
-        type ``[,]``<'T> with
```
Comment (TIHan) [inline src/fsharp/FSharp.Core/prim-types.fs]: To note: We determined this was not used anywhere and is also not public. Effectively, dead code.

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 2]
```diff
@@ -651,6 +651,12 @@ namespace Microsoft.FSharp.Core
             let inline GetArray2DLength1 (arr: 'T[,]) =  (# "ldlen.multi 2 0" arr : int #)  
             let inline GetArray2DLength2 (arr: 'T[,]) =  (# "ldlen.multi 2 1" arr : int #)  
 
+            let inline GetArray2DLength (arr: 'T[,]) (dim: int) = 
+                match dim with
+                    | 0 -> GetArray2DLength1 arr
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fs]: Style nit: align with `|` with the `m` in `match`

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 3]
```diff
@@ -679,6 +685,13 @@ namespace Microsoft.FSharp.Core
 
             let inline GetArray3DLength3 (arr: 'T[,,]) =  (# "ldlen.multi 3 2" arr : int #)  
 
+            let inline GetArray3DLength (arr: 'T[,,]) (dim: int) = 
+                match dim with
+                    | 0 -> GetArray3DLength1 arr
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fs]: Same style nit

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 4]
```diff
@@ -4894,39 +4907,39 @@ namespace Microsoft.FSharp.Core
...
-                for j = 0 to len2 - 1 do 
-                    SetArray dst j (GetArray2D source index1 (start2+j))
+            let inline GetArraySlice2DFixed (source: _[,]) start finish index nonFixedDim = 
+                let bound = source.GetLowerBound(nonFixedDim)
+                let start, finish = ComputeSlice bound start finish (GetArray2DLength source nonFixedDim)
+                let len = (finish - start + 1)
+                let dst = zeroCreate (if len < 0 then 0 else len)
+                let getArrayElem = 
+                    match nonFixedDim with
+                        | 1 -> (fun i -> GetArray2D source index (start+i))
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fs]: Same style nit

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 5]
```diff
@@ -4894,39 +4907,39 @@ namespace Microsoft.FSharp.Core
...
-                let len1 = (finish1 - start1 + 1)
-                for i = 0 to len1 - 1 do
-                    SetArray2D target (bound1+start1+i) index2 (GetArray source i)
+            let inline SetArraySlice2DFixed (target: _[,]) (source: _[]) index start finish nonFixedDim  = 
+                let bound = target.GetLowerBound(nonFixedDim)
+                let start, finish = ComputeSlice bound start finish (GetArray2DLength target nonFixedDim)
+                let len = (finish - start + 1)
+                let setArrayElem = 
+                    match nonFixedDim with
+                        | 1 -> (fun j -> SetArray2D target index (bound + start + j) (GetArray source j))
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fs]: Same style nit

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 6]
```diff
@@ -4949,6 +4962,55 @@ namespace Microsoft.FSharp.Core
...
+                let bound2 = source.GetLowerBound(nonFixedDim2)
+                let start1, finish1 = ComputeSlice bound1 start1 finish1 (GetArray3DLength source nonFixedDim1)              
+                let start2, finish2 = ComputeSlice bound2 start2 finish2 (GetArray3DLength source nonFixedDim2)              
+                let len1 = (finish1 - start1 + 1)
+                let len2 = (finish2 - start2 + 1)
+
+                let dst = Array2DZeroCreate (max 0 len1) (max 0 len2)
+                let getArrayElem = 
+                    match nonFixedDim1, nonFixedDim2 with
+                        | 1,2 -> (fun i j -> GetArray3D source index (start1 + i) (start2 + j))
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fs]: Same style nit

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 7]
```diff
@@ -4949,6 +4962,55 @@ namespace Microsoft.FSharp.Core
...
+            let inline GetArraySlice3DFixedSingle3 (source: _[,,]) start1 finish1 start2 finish2 index3 = GetArraySlice3DFixedSingle source start1 finish1 start2 finish2 index3 0 1
+
+            let inline GetArraySlice3DFixedDouble (source: _[,,]) start finish index1 index2 nonFixedDim = 
+                let bound = source.GetLowerBound(nonFixedDim)
+                let start, finish = ComputeSlice bound start finish (GetArray3DLength source nonFixedDim)
+                let len = (finish - start + 1)
+                let dst = zeroCreate (if len < 0 then 0 else len)
+                let getArrayElem = 
+                    match nonFixedDim with
+                        | 2 -> (fun i -> GetArray3D source index1 index2 i)
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fs]: Same style nit

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 8]
```diff
@@ -4961,6 +5023,52 @@ namespace Microsoft.FSharp.Core
...
+            let inline SetArraySlice3DFixedSingle (target: _[,,]) (source: _[,]) index start1 finish1 start2 finish2 nonFixedDim1 nonFixedDim2  = 
+                let bound1 = target.GetLowerBound(nonFixedDim1)
+                let bound2 = target.GetLowerBound(nonFixedDim2)
+                let start1, finish1 = ComputeSlice bound1 start1 finish1 (GetArray3DLength target nonFixedDim1)              
+                let start2, finish2 = ComputeSlice bound2 start2 finish2 (GetArray3DLength target nonFixedDim2)              
+                let len1 = (finish1 - start1 + 1)
+                let len2 = (finish2 - start2 + 1)
+                let setArrayElem = 
+                    match nonFixedDim1, nonFixedDim2 with
+                        | 1,2 -> (fun i j -> SetArray3D target index (bound1 + start1 + i) (bound2 + start2 + j) (GetArray2D source i j))
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fs]: Same style nit, also consider `1, 2` instead of `1,2`

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 9]
```diff
@@ -4961,6 +5023,52 @@ namespace Microsoft.FSharp.Core
...
+            let inline SetArraySlice3DFixedSingle3 (target: _[,,]) start1 finish1 start2 finish2 index (source: _[,]) = 
+                SetArraySlice3DFixedSingle target source index start1 finish1 start2 finish2 0 1
+
+            let inline SetArraySlice3DFixedDouble (target: _[,,]) (source: _[]) index1 index2 start finish nonFixedDim = 
+                let bound = target.GetLowerBound(nonFixedDim)
+                let start, finish = ComputeSlice bound start finish (GetArray3DLength target nonFixedDim)
+                let len = (finish - start + 1)
+                let setArrayElem = 
+                    match nonFixedDim with
+                        | 2 -> (fun k -> SetArray3D target index1 index2 (bound + start + k) (GetArray source k))
```
Comment (cartermp) [inline src/fsharp/FSharp.Core/prim-types.fs]: Same style nit

**Comment by cartermp** [comment]
@dsyme thoughts on the implementation?

**Comment by TIHan** [comment]
We were talking about this yesterday. We may or may not have decided to take another approach where we would use an extension, `type ``[,,]`` with` to provide the overloads at that level which would not require TypeChecker changes and those overloads would not be marked `inline`, effectively removing the need for the intrinsic functions to be public. Our concern was that we would be adding 20 intrinsic APIs to the public surface area (if we did the same for Array4D).

However, after looking at this from the PR's perspective, we probably should continue following the current pattern of the intrinsic public functions. It may not be good to suddenly add `GetSlice` extension methods where the other arrays did not have such a thing and only relied on intrinsics. As long as we follow the existing pattern, that is fine.  If we decide that we want to support more than array4d, when that time comes, it might be valid to take a different path. In this case, Array3D and Array4D already exist and should be a parity with others, at feature and public API level.

Therefore, I think this is fine too. Great work @nelson-wu .

**Pull request merged**